PSPs/drafts/psp-30.md

```diff
@@ -0,0 +1,882 @@
+Selector: `` - first 4 bytes of `blake2b_256("")` // TO UPDATE WHEN PSP NUMBER
+```json
+{
+  "args": [
+    {
+      "name": "operator",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "name": "from",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "name": "id",
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    },
+    {
+      "name": "data",
+      "type": {
+        "displayName": [
+          "[u8]"
+        ],
+        "type": "[u8]"
+      }
+    }
+  ],
+  "docs": [
+    " Ensures that the smart contract allows reception of PSP30 token(s).",
+    " Returns `Ok(())` if the contract allows the reception of the token(s) and Error `TransferRejected(String))` otherwise.",
+    "",
+    " This method will get called on every transfer to check whether the recipient in `transfer`",
+    " or `transfer_from` is a contract, and if it is, does it accept tokens.",
+    " This is done to prevent contracts from locking tokens forever.",
+    "",
+    " Returns `PSP30ReceiverError` if the contract does not accept the tokens."
+  ],
+  "mutates": true,
+  "name": [
+    "PSP30Receiver",
+    "before_received"
+  ],
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "Result"
+    ],
+    "type": 2
+  },
+  "selector": "" // TO UPDATE WHEN PSP NUMBER
+}
+```
+
+### Extension
+
+#### PSP30Metadata
+
+`PSP30Metadata` is an optional interface for metadata for this Non-Fungible Token standard.
+
+##### **name**() ➔ Option<String>
+Selector: `` - first 4 bytes of `blake2b_256("")` // TO UPDATE WHEN PSP NUMBER
+```json
+{
+  "args": [],
+  "docs": [
+    "Returns the token name."
+  ],
+  "mutates": false,
+  "name": [
+    "PSP30Metadata",
+    "name"
+  ],
+  "returnType": {
+    "displayName": [
+      "Option"
+    ],
+    "type": "Option<string>"
+  },
+  "selector": "" // TO UPDATE WHEN PSP NUMBER
+}
+```
+
+##### **symbol**() ➔ Option<String>
+Selector: `` - first 4 bytes of `blake2b_256("")` // TO UPDATE WHEN PSP NUMBER
+```json
+{
+  "args": [],
+  "docs": [
+    "Returns the token symbol."
+  ],
+  "mutates": false,
+  "name": [
+    "PSP22Metadata",
+    "symbol"
+  ],
+  "returnType": {
+    "displayName": [
+      "Option"
+    ],
+    "type": "Option<string>"
+  },
+  "selector": "" // TO UPDATE WHEN PSP NUMBER
+}
+```
+
+##### **uri**(id: Id) ➔ Option<String>
+Selector: `` - first 4 bytes of `blake2b_256("")` // TO UPDATE WHEN PSP NUMBER
+```json
+{
+  "args": [
+    {
+      "name": "id",
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    }
+  ],
+  "docs": [
+    " Returns the Uniform Resource Identifier (URI) for `id` token."
```

Comment:
I think we should describe that if the token is invalid, it should return `None`.

Also we need to define what exactly should be returned.

Initial Erc721 returns a link on `json` that has schema defiend [here](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md#specification)(ERC721 Metadata JSON Schema section).

But I read several ideas regarding NFTs and seems that they require additional fields.

So I think the final schema can be something like this:
```json
{
    "title": "Asset Metadata",
    "type": "object",
    "properties": {
        "name": {
            "type": "string",
            "description": "Identifies the asset to which this NFT represents",
        },
        "description": {
            "type": "string",
            "description": "Describes the asset to which this NFT represents",
        },
        "image": {
            "type": "string",
            "description": "A URI pointing to a resource with mime type image/* representing the asset to which this NFT represents. Consider making any images at a width between 320 and 1080 pixels and aspect ratio between 1.91:1 and 4:5 inclusive.",
        },
        "external_uri": {
            "type": "string",
            "description": "A URI pointing to a a externally hosted resources, usually the originating market place. Allows for 3rd party services to link to the originally NFT.",
            "optional": true
        },
        "attributes": {
            "type": "object",
            "description": "Defines a set of attributes which classifies and defines the NFT. Allows for applications to build fitlers and attribute search functionalities for a set of NFTs.",
            "optional": true
        }
    }
}
```